go/libraries/doltcore/table/editor/pk_table_editor.go

```diff
@@ -125,16 +122,19 @@ type pkTableEditor struct {
 	tea      TableEditAccumulator
 	nbf      *types.NomsBinFormat
 	indexEds []*IndexEditor
-	cvEditor *types.MapEditor
 
 	// TupleFactory is not thread-safe.  writeMutex needs to be acquired before using tf
 	tf *types.TupleFactory
 
 	// Whenever any write operation occurs on the table editor, this is set to true for the lifetime of the editor.
 	dirty uint32
 
-	// This mutex blocks on each operation, so that map reads and updates are serialized
+	// This mutex blocks on each operation, so that map reads and updates are serialized. Protects the above data.
 	writeMutex *sync.Mutex
+
+	// protects cvEditor
+	cvMutex  *sync.Mutex
```

Comment:
I created a new mutex to avoid a deadlock:
1. write lock is taken on `writeMutex` when InsertRow is called
2. Any duplicate rows are passed to `PkDuplicateCb` which calls `SetConstraintViolation` which tries to take a lock.
3. Deadlock.